vyper/codegen/module.py

```diff
@@ -47,92 +50,357 @@ def _is_payable(func_ast):
-        func_ir = generate_ir_for_function(func_ast, global_ctx, skip_nonpayable_check)
-        selector_section.append(func_ir)
+            function_infos.extend([method_id_bytes, symbol, func_metadata])
 
-    if default_function:
-        fallback_ir = generate_ir_for_function(
-            default_function, global_ctx, skip_nonpayable_check=False
-        )
-    else:
-        fallback_ir = IRnode.from_list(
-            ["revert", 0, 0], annotation="Default function", error_msg="fallback function"
-        )
+        selector_section.append(function_infos)
 
-    # ensure the external jumptable section gets closed out
-    # (for basic block hygiene and also for zksync interpreter)
-    # NOTE: this jump gets optimized out in assembly since the
-    # fallback label is the immediate next instruction,
-    close_selector_section = ["goto", "fallback"]
+    ret = ["seq", ["with", "_calldata_method_id", shr(224, ["calldataload", 0]), selector_section]]
 
-    global_calldatasize_check = ["if", ["lt", "calldatasize", 4], ["goto", "fallback"]]
+    ret.extend(function_irs)
 
-    runtime = [
-        "seq",
-        global_calldatasize_check,
-        ["with", "_calldata_method_id", shr(224, ["calldataload", 0]), selector_section],
-        close_selector_section,
-        ["label", "fallback", ["var_list"], fallback_ir],
-    ]
+    return ret
 
-    runtime.extend(internal_functions_ir)
 
-    return runtime
+# codegen for all runtime functions + callvalue/calldata checks,
+# with O(1) jumptable for selector table.
+# uses two level strategy: uses `method_id % n_methods` to calculate
+# a bucket, and then descends into linear search from there.
+# costs about 126 gas for typical (nonpayable, >0 args, avg bucket size 1.5)
+# function and 24 bytes of code (+ ~23 bytes of global overhead)
+def _selector_section_sparse(external_functions, global_ctx):
+    ret = ["seq"]
+
+    if len(external_functions) == 0:
+        return ret
+
+    entry_points, sig_of = _generate_external_entry_points(external_functions, global_ctx)
+
+    n_buckets, buckets = jumptable_utils.generate_sparse_jumptable_buckets(entry_points.keys())
+
+    # 2 bytes for bucket location
+    SZ_BUCKET_HEADER = 2
+
+    if n_buckets > 1:
+        bucket_id = ["mod", "_calldata_method_id", n_buckets]
+        bucket_hdr_location = [
+            "add",
+            ["symbol", "selector_buckets"],
+            ["mul", bucket_id, SZ_BUCKET_HEADER],
+        ]
+        # get bucket header
+        dst = 32 - SZ_BUCKET_HEADER
+        assert dst >= 0
+
+        if _is_debug_mode():
+            ret.append(["assert", ["eq", "msize", 0]])
+
+        ret.append(["codecopy", dst, bucket_hdr_location, SZ_BUCKET_HEADER])
+
+        jumpdest = IRnode.from_list(["mload", 0])
+        # don't particularly like using `jump` here since it can cause
+        # issues for other backends, consider changing `goto` to allow
+        # dynamic jumps, or adding some kind of jumptable instruction
```

Comment:
so we don't optimise here for pluggable backends? why exactly since this can be of value-add...

Reply:
the reason here is that in order to get a proper looking indirect jump instruction we would need the data to be packed differently which would result in more CODECOPY instructions than we need. however, i specifically implemented a linear search selector table in `--optimize none` mode so other backends still work (and confirmed with zksync)